server/src/main/java/org/elasticsearch/ingest/IngestMetric.java

```diff
@@ -53,7 +57,19 @@ void preIngest() {
      */
     void postIngest(long ingestTimeInNanos) {
         long current = ingestCurrent.decrementAndGet();
-        assert current >= 0 : "ingest metric current count double-decremented";
+        if (current < 0) {
+            /*
+             * This ought to never happen. However if it does, it's incredibly bad because ingestCurrent being negative causes a
+             * serialization error that prevents the nodes stats API from working. So we're doing 3 things here:
+             * (1) Log a stack trace at warn level so that the Elasticsearch engineering team can track down and fix the source of the
+             * bug if it still exists
+             * (2) Throw an AssertionError if assertions are enabled so that we are aware of the bug
+             * (3) Increment the counter back up so that we don't hit serialization failures
+             */
+            logger.warn("Current ingest counter decremented below 0", new RuntimeException());
+            assert false : "ingest metric current count double-decremented";
+            ingestCurrent.incrementAndGet();
```

Comment:
This doesn't _quite_ fix the serialization problem because `createStats()` might capture a negative value for `ingestCurrent` if it runs just before this line. I mean it won't be a permanent failure any more but still I think we should have a `max(0,...)` in `createStats()` too.